### ISSUE TITLE
REQUEST FOR HELP WITH APPLYING TRANSFER LEARNING ON YOLO11 (FREEZE BACKBONE + NECK / UNFREEZE HEAD (DETECT) TO TRAIN)

### ISSUE THREAD

**Issue by sonnam2003** [ultralytics/ultralytics#22380]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Other

### Bug

Dear **Ultralytics Team**,

I am currently a final year student studying the topic Using YOLO11 to detect real time marine debris. Currently, I am facing an issue where I am unsure whether I am correctly applying TL (Transfer Learning). I genuinely do not know the actual state of my model — whether TF has been properly applied or not. Specifically, I have tried training the model with and without applying TL, but both approaches yield identical results in terms of performance and training time. According to theory, however, the outcomes should differ significantly.

Here is my process of applying TL to pretrained model YOLO11:

**1. Firstly, I loaded the pre-trained model YOLO11s and show status of all layers:**

> from ultralytics import YOLO
import torch
model = YOLO("yolo11s.pt")
for name, param in model.model.named_parameters():
    status = "Unfrozen" if param.requires_grad else " Frozen"
    print(f"{name:<60} {status}")

<img width="1056" height="835" alt="Image" src="https://github.com/user-attachments/assets/63be9b9d-7bcf-42f6-b270-f9f1d1b68969" />

**2. Then I set the layer that I would like to unfreeze to train on my dataset**

`for p in model.model.parameters():
    p.requires_grad = True

for name, param in model.model.named_parameters():
    if not any(x in name for x in [ "model.20","model.22","model.23"]):
        param.requires_grad = False`

**3.  Then I checked again to confirm "model.20, 22, 23" has been unfrozen to train by using this code:**

> print("\n Status của model sau khi FREEZE / UNFREEZE thủ công")
trainable_params = 0
frozen_params = 0
for name, param in model.model.named_parameters():
    if param.requires_grad:
        status = " 🔴 TRAINING"
        trainable_params += param.numel()
    else:
        status = " 🟢 FROZEN"
        frozen_params += param.numel()
    print(f"{status:<12} | {name:<60} | {param.numel():>10,} params")
print("\n ------ Conclusion ------")
print(f"🟢 Trainable parameters : {trainable_params:,}")
print(f"🔴 Frozen parameters    : {frozen_params:,}")
print(f" Total parameters           : {trainable_params + frozen_params:,}")

<img width="723" height="870" alt="Image" src="https://github.com/user-attachments/assets/7e930dad-0ccd-4302-94ab-ccba8a960576" />

**4.  Then I confirmed once again by checking the optimize parameter (= trainable parameter)**

<img width="824" height="383" alt="Image" src="https://github.com/user-attachments/assets/484de8e9-38e5-4aef-9d7b-ab2dbd9e8e05" />

**5. Finally, I started train the model**

<img width="541" height="217" alt="Image" src="https://github.com/user-attachments/assets/fc48950e-b2d6-46ea-949a-9428027b8ad0" />

**6. Result of some of the first epochs**

<img width="1383" height="537" alt="Image" src="https://github.com/user-attachments/assets/a2f80e99-0072-4aa5-8beb-5b34f1c5b4fa" />

-----------

After tried with transfer learning like the above process, I decided to try train again without applying Transfer learning by unfreeze all layers from model.0 to model.23 to check the performance is difference or not, with the similar process , just different at the stage that I adjust to unfreeze all the layer:

<img width="547" height="154" alt="Image" src="https://github.com/user-attachments/assets/a59729fd-2bbf-49e8-ae05-b7782494ebdb" />

<img width="734" height="739" alt="Image" src="https://github.com/user-attachments/assets/6a6988af-ac70-4c04-9aa2-a5f0b791f185" />

But the result was exactly the same to the way I apply transfer learning, because as I know that if we train model without apply transfer learning, this mean the model will train again all the parameter from the start, which really takes more lots of time than train model with apply transfer learning. 

<img width="1337" height="599" alt="Image" src="https://github.com/user-attachments/assets/08d4541f-a9a9-4b32-ba90-c886b1b11501" />

(As we can see that epoch 1 has mAP50  = 0.202 exactly the same with result of applying TL, similar to other epochs)

Therefore, I kindly request assistance and detailed guidance on whether my method of applying for Transfer learning is correct. If it is not, I would appreciate a thorough explanation, as I am a student currently learning. Additionally, could you please inform me of the reasons why I am experiencing the issues mentioned above?

Thanks and best regards,
Nam

### Environment

Ultralytics 8.3.207 🚀 Python-3.11.13 torch-2.6.0+cu124 CUDA:0 (Tesla T4, 15095MiB)
Setup complete ✅ (4 CPUs, 31.4 GB RAM, 6520.0/8062.4 GB disk)

OS                     Linux-6.6.56+-x86_64-with-glibc2.35
Environment            Colab
Python                 3.11.13
Install                pip
Path                   /usr/local/lib/python3.11/dist-packages/ultralytics
RAM                    31.35 GB
Disk                   6520.0/8062.4 GB
CPU                    Intel Xeon CPU @ 2.00GHz
CPU count              4
GPU                    Tesla T4, 15095MiB
GPU count              2
CUDA                   12.4

numpy                  ✅ 1.26.4>=1.23.0
matplotlib             ✅ 3.7.2>=3.3.0
opencv-python          ✅ 4.11.0.86>=4.6.0
pillow                 ✅ 11.3.0>=7.1.2
pyyaml                 ✅ 6.0.3>=5.3.1
requests               ✅ 2.32.5>=2.23.0
scipy                  ✅ 1.13.0>=1.4.1
torch                  ✅ 2.6.0+cu124>=1.8.0
torch                  ✅ 2.6.0+cu124!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision            ✅ 0.21.0+cu124>=0.9.0
psutil                 ✅ 7.1.0
polars                 ✅ 1.25.0
ultralytics-thop       ✅ 2.0.17>=2.0.0
{'OS': 'Linux-6.6.56+-x86_64-with-glibc2.35',
 'Environment': 'Colab',
 'Python': '3.11.13',
 'Install': 'pip',
 'Path': '/usr/local/lib/python3.11/dist-packages/ultralytics',
 'RAM': '31.35 GB',
 'Disk': '6520.0/8062.4 GB',
 'CPU': 'Intel Xeon CPU @ 2.00GHz',
 'CPU count': 4,
 'GPU': 'Tesla T4, 15095MiB',
 'GPU count': 2,
 'CUDA': '12.4',
 'Package Info': {'numpy': '✅ 1.26.4>=1.23.0',
  'matplotlib': '✅ 3.7.2>=3.3.0',
  'opencv-python': '✅ 4.11.0.86>=4.6.0',
  'pillow': '✅ 11.3.0>=7.1.2',
  'pyyaml': '✅ 6.0.3>=5.3.1',
  'requests': '✅ 2.32.5>=2.23.0',
  'scipy': '✅ 1.13.0>=1.4.1',
  'torch': '✅ 2.6.0+cu124!=2.4.0,>=1.8.0; sys_platform == "win32"',
  'torchvision': '✅ 0.21.0+cu124>=0.9.0',
  'psutil': '✅ 7.1.0',
  'polars': '✅ 1.25.0',
  'ultralytics-thop': '✅ 2.0.17>=2.0.0'}}

### Minimal Reproducible Example

from ultralytics import YOLO
import torch

model = YOLO("yolo11s.pt")

for name, param in model.model.named_parameters():
    status = "Unfrozen" if param.requires_grad else " Frozen"
    print(f"{name:<60} {status}")

for p in model.model.parameters():
    p.requires_grad = True

for name, param in model.model.named_parameters():
    if not any(x in name for x in [ "model.20","model.22","model.23"]):
        param.requires_grad = False

print("\n Status của model sau khi FREEZE / UNFREEZE thủ công")
trainable_params = 0
frozen_params = 0

for name, param in model.model.named_parameters():
    if param.requires_grad:
        status = " 🔴 TRAINING"
        trainable_params += param.numel()
    else:
        status = " 🟢 FROZEN"
        frozen_params += param.numel()
    
    print(f"{status:<12} | {name:<60} | {param.numel():>10,} params")

print("\n ------ Conclusion ------")
print(f"🟢 Trainable parameters : {trainable_params:,}")
print(f"🔴 Frozen parameters    : {frozen_params:,}")
print(f" Total parameters           : {trainable_params + frozen_params:,}")

optimizer = torch.optim.SGD([p for p in model.parameters() if p.requires_grad], lr=1e-3)
print(f"The number of optimizer parameter will update: {sum(p.numel() for p in optimizer.param_groups[0]['params']):,}")

### Are you willing to submit a PR?

- [x] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @sonnam2003, thank you for your interest in Ultralytics 🚀! This is an automated response to help gather the right details quickly — an Ultralytics engineer will also assist you here soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

To help us investigate your YOLO11 transfer learning scenario, please also include:
-Exact training command or Python snippet used to start training (args like epochs, imgsz, batch, lr, freeze, device)  
-Model variant and weights used (e.g., yolo11s.pt)  
-Data config YAML and number of classes  
-Console logs from the first 3–5 epochs and final metrics  
-Confirmation of which layers you intended to freeze/unfreeze and how you verified it  
-A short note on dataset size and class balance, plus a few representative images

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by sonnam2003** [comment]
Dear **Mr. @glenn-jocher** and **Ultralytics team**,

I would greatly appreciate your assistance regarding the issue I mentioned above. Please kindly review it, and if you require any additional information or clarification, do not hesitate to let me know.

Thank you very much for your time and support.

**Comment by Y-T-G** [comment]
It's using transfer learning as long as you are training from `.pt` model. You don't need to do anything else. It's not training from scratch.

Keeping all the layers unfrozen, or partially freezing them doesn't change that it's transfer learning. Transfer learning is not connected to that. It's transfer learning as long as you start from a pretrained checkpoint. In this case, you are starting from the `yolo11s.pt` pretrained checkpoint.
